lxd/storage/drivers/driver_ceph_volumes.go

```diff
@@ -327,6 +327,16 @@ func (d *ceph) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots boo
 	// Function to run once the volume is created, which will regenerate the filesystem UUID (if needed),
 	// ensure permissions on mount path inside the volume are correct, and resize the volume to specified size.
 	postCreateTasks := func(v Volume) error {
+		// For VMs, also copy the filesystem volume.
+		if vol.IsVMBlock() {
+			srcFSVol := srcVol.NewVMBlockFilesystemVolume()
+			fsVol := vol.NewVMBlockFilesystemVolume()
+			err := d.CreateVolumeFromCopy(fsVol, srcFSVol, false, false, op)
```

Comment:
> At least for the Ceph driver `DeleteVolume()` doesn't check if the volume has snapshots. So we are safe calling it using the reverter in case of error.

Yes but we would want to change that so its consistent with the others.

Reply:
> Contrary to the comment in the ZFS driver, down the line it calls `deleteDatasetRecursive()` which runs `zfs destroy -r` which should also delete the volumes snapshots. Wouldn't it be a breaking change if we add the error in case of snapshots to `DeleteVolume()` of the ZFS driver?

Indeed, we would want to change that so its consistent with the other drivers' public interface behavior.

We can then either add a `deleteSnapshots bool` to the public `DeleteVolume()` interface and all the driver implementations and update the call sites as needed. Or implement an internal recursive delete function for ZFS (although I would question why only ZFS needs that first to check we're not missing something).

Reply:
> LVM isn't currently doing the revert of the filesystem volume either, for Dir and Btrfs it's not clear to me yet.

Indeed they should do so too.

Reply:
What is now even more confusing, none of the checks even has any effect. Let's take dir for an example:

```
lxc storage create dir dir
lxc storage volume create dir vol size=10MiB
lxc storage volume snapshot dir vol
lxc storage volume rm dir vol
```

The above succeeds for at least Dir and Btrfs.

Reply:
No, I suspect its either LXD itself or the backendLXD that is clearing the snapshots first.
But the driver interface and the drive behavior should be consistent.

Reply:
Ah ok, that makes sense yes. Then let me add in the error for Ceph and we can merge this PR.